### ISSUE TITLE
[surface] MLS VOXEL_GRID_DILATION upsampling method gives extremely large point cloud

### ISSUE THREAD

**Issue by keineahnung2345** [PointCloudLibrary/pcl#4945]
**Describe the bug**

MLS `VOXEL_GRID_DILATION` upsampling method gives extremely large point cloud.

**Context**

I am testing the code posted at https://github.com/PointCloudLibrary/pcl/issues/1958 (I put my revision [here](https://gist.github.com/keineahnung2345/f3f323e4c643770036fcdf59a648e007)) and the upsampling result of `VOXEL_GRID_DILATION` seems weird to me. I am playing with the parameter dilation iteration, when it's set to 0, the result is:

![image](https://user-images.githubusercontent.com/18047300/133444598-d4509a17-24f4-4d31-a386-648ebe620f49.png)

When it's set to 1, the result becomes:

![image](https://user-images.githubusercontent.com/18047300/133444731-e1d3254f-ce98-4108-8396-d3f2f6b2f90c.png)

look it in larger scale:

![image](https://user-images.githubusercontent.com/18047300/133444873-5b19af5b-f37c-43b1-8703-f1c3366479d3.png)

**Current Behavior**

The point cloud dimension becomes around 1e+7 when dilation iteration is set to 1, for comparison, the dimension is around 10 when dilation iteration is set to 0. 

**Expected behavior**

What I expect is that the the point cloud dimension when dilation iteration is set to 1 should be near to that when dilation iteration is set to 0.

**To Reproduce**

https://gist.github.com/keineahnung2345/f3f323e4c643770036fcdf59a648e007

**Your Environment (please complete the following information):**

 - OS: Ubuntu 20.04
 - PCL Version: master branch

**Possible Solution**

https://github.com/PointCloudLibrary/pcl/blob/93d04f8b8b3408b0db4d823a249d03b8547b48d7/surface/include/pcl/surface/impl/mls.hpp#L873-L886

`new_index` may contain negative element if `index` is (0,0,0) and (x,y,z) is (-1,-1,-1), in this case, the `std::uint64_t index_1d` will overflow and I guess it's the cause of this strange behavior. The solutions I come up is either:
1. allow the key of `pcl::MovingLeastSquares::MLSVoxelGrid::HashMap` to be negative and rewrite `getIndexIn1D` and `getIndexIn3D`
2. while dilating, don't allow the voxel grids to exceed the bounding box of the point cloud

Which one is better?

**Comment by mvieth** [comment]
Hmm, the second solution does not seem satisfactory.
@tin1254 What are your thoughts about this?

**Comment by tin1254** [comment]
can't think of any better solution except changing the hash map in `MLSVoxelGrid`. I think it's good enough as a quick remedy. @mvieth

**Comment by mvieth** [comment]
In which way would you change the hash map?
I think not only negative indices are a problem, see for example this 2D equivalent (bounding box in blue):
![MLSVoxelGrid](https://user-images.githubusercontent.com/39675748/134315702-55e19a29-b519-4127-b088-91745e393c0c.png)
On the lower left, the three indices 3, 7, 11 are not negative, but would refer to cells on the far right.
So a solution could be to change the bounding box to a bigger one in each dilation step, or keep the bounding box and forbid dilation outside of the box, as @keineahnung2345 suggested
BTW this is related to [pairing functions](https://en.wikipedia.org/wiki/Pairing_function), but they don't seem to be applicable here since they deal with natural numbers, not integer numbers (positive+negative numbers).

**Comment by tin1254** [comment]
I would suggest increasing the value of `bounding_min_`, `bounding_max_` according to `dilation_iteration_num_` and `voxel_size_`. In that case, we could skip checking for each point, because we already know where will the maximum and minimum point(s) locate given those two values.

https://github.com/PointCloudLibrary/pcl/blob/c9dd65bfd829618546eea826d80fb2751e9928c5/surface/include/pcl/surface/mls.h#L616-L621

For example in 2D case, if we have `dilation_iteration_num_ = 2`, then we decrease the value of `bounding_min_` to make the minimum point in the original cloud `getCellIndex(min_p) = [2, 2]` (voxel idx). After two dilations, the minimum points in the new cloud will be `[0, 2]` and `[2, 0]`, which are still inside the bound.

What do you think? @mvieth

**Comment by mvieth** [comment]
If `bounding_min_` and `bounding_max` are adjusted, then `data_size_` also has to be adjusted, and all indices in the hashmap recomputed, right?
The alternative would be to check if all entries of `new_index` are non-negative [here](https://github.com/PointCloudLibrary/pcl/blob/master/surface/include/pcl/surface/impl/mls.hpp#L879), and to otherwise not dilate in that direction. This would have the effect that the dilation only happens inside the original bbox (not sure how bad that is in this scenario), but this fix would be easier and less error prone.
If you like and have the time, maybe create a pull request with whatever solution seems best to you?

**Comment by tin1254** [comment]
> and all indices in the hashmap recomputed, right?

Maybe my explanation wasn't so clear before. My suggestion was adjusting `bounding_min_` and `bounding_max_` at the beginning of constructing the hash map with the original cloud, so indices of the voxel won't be rehashed because all the voxels (including the dilated ones) are always inside the bound

> The alternative would be to check if all entries of new_index are non-negative here

NP, I can try both solutions and create a PR in a week or two

**Comment by keineahnung2345** [comment]
> BTW this is related to pairing functions, but they don't seem to be applicable here since they deal with natural numbers, not integer numbers (positive+negative numbers).

Hi, I have found some variant of cantor pairing function that works for integer and also for three numbers, and I put my python implementation [here](https://gist.github.com/keineahnung2345/8b2ea500e7f56a4a1a8fee7dcc03ce52). I've tested it in a very limited range(a,b,c in [-3,3]), and the results seem to be correct.

I think it should be helpful here, but I haven't gone through the proof yet, so its correctness in a larger range is to be confirmed.
